wkcuber/export_wkw_as_tiff.py

```diff
@@ -0,0 +1,218 @@
+
+
+def wkw_slice_to_image(data_slice: np.ndarray) -> Image:
+    # discard the z dimension
+    data_slice = data_slice.squeeze(axis=3)
+    if data_slice.shape[0] == 1:
+        # discard greyscale dimension
+        data_slice = data_slice.squeeze(axis=0)
+        # swap the axis
+        data_slice = data_slice.transpose((1, 0))
+    else:
+        # swap axis and move the channel axis
+        data_slice = data_slice.transpose((2, 1, 0))
+    return Image.fromarray(data_slice)
+
+
+def export_tiff_slice(
+    export_args: Tuple[
+        int,
+        Tuple[
+            Dict[str, Tuple[int, int, int]], str, str, str, Union[None, Tuple[int, int]]
+        ],
+    ]
+):
+    slice_number, (tiff_bbox, dest_path, name, dataset_path, tiling_size) = export_args
+    tiff_bbox["topleft"] = [
+        tiff_bbox["topleft"][0],
+        tiff_bbox["topleft"][1],
+        tiff_bbox["topleft"][2] + slice_number - 1,
+    ]
+    tiff_bbox["size"] = [tiff_bbox["size"][0], tiff_bbox["size"][1], 1]
+
+    with wkw.Dataset.open(dataset_path) as dataset:
+        if tiling_size is None:
+            tiff_data = dataset.read(tiff_bbox["topleft"], tiff_bbox["size"])
+            tiff_file_name = wkw_name_and_bbox_to_tiff_name(name, slice_number)
+
+            tiff_file_path = os.path.join(dest_path, tiff_file_name)
+
+            image = wkw_slice_to_image(tiff_data)
+            image.save(tiff_file_path)
+            logging.info(f"saved slice {slice_number}")
+
+        else:
+            tile_bbox_size = [tiling_size[0], tiling_size[1], 1]
+            for y_tile_index in range(ceil(tiff_bbox["size"][1] / tiling_size[1])):
+                tile_tiff_path = os.path.join(
+                    dest_path, str(slice_number), str(y_tile_index + 1)
+                )
+                os.makedirs(tile_tiff_path, exist_ok=True)
+                for x_tile_index in range(ceil(tiff_bbox["size"][0] / tiling_size[0])):
+                    tile_tiff_filename = f"{x_tile_index + 1}.tiff"
+                    tile_bbox_topleft = [
+                        tiff_bbox["topleft"][0] + tile_bbox_size[0] * x_tile_index,
+                        tiff_bbox["topleft"][1] + tile_bbox_size[1] * y_tile_index,
+                        tiff_bbox["topleft"][2],
+                    ]
+                    tile_tiff_data = dataset.read(
+                        off=tile_bbox_topleft, shape=tile_bbox_size
+                    )
+                    tile_image = wkw_slice_to_image(tile_tiff_data)
+                    tile_image.save(os.path.join(tile_tiff_path, tile_tiff_filename))
+
+    logging.info(f"saved all tiles of slice {slice_number}")
+
+
+def export_tiff_stack(
+    wkw_file_path, wkw_layer, bbox, mag, destination_path, name, tiling_slice_size, args
+):
+    os.makedirs(destination_path, exist_ok=True)
+    dataset_path = os.path.join(wkw_file_path, wkw_layer, mag.to_layer_name())
+
+    with get_executor_for_args(args) as executor:
+        num_slices = bbox["size"][2]
```

Comment:
I think it would make sense to iterate over batches of slices. Since wkw is organized in 32**3 cubes, 32 slices have to be read even if you're only interested in a single slice. Therefore, it is more efficient to read in 32 slices at once and write them out one after the other. The batch size should be 32 by default and adjustable through args. See `wkcuber.cubing` for reference.